transport/http/agent_test.go

```diff
@@ -0,0 +1,389 @@
+type peerIdentifierMatcher hostport.PeerIdentifier
+
+// Matches returns whether x is a match.
+func (pim peerIdentifierMatcher) Matches(x interface{}) bool {
+	res, ok := x.(transport.PeerIdentifier)
+	if !ok {
+		return false
+	}
+
+	return res.Identifier() == hostport.PeerIdentifier(pim).Identifier()
+}
+
+// String describes what the matcher matches.
+func (pim peerIdentifierMatcher) String() string {
+	return string(hostport.PeerIdentifier(pim).Identifier())
+}
+
+func TestAgent(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	type testStruct struct {
+		msg           string
+		agent         *Agent
+		appliedFunc   func(*Agent) error
+		expectedPeers []peerExpectation
+		expectedErr   error
+	}
+	tests := []testStruct{
+		func() (s testStruct) {
+			s.msg = "one retain"
+			s.agent = NewDefaultAgent()
+			s.expectedPeers = createPeerExpectations(
+				mockCtrl,
+				[]string{"localhost:1234"},
+				1,
+			)
+			s.appliedFunc = func(a *Agent) error {
+				_, err := a.RetainPeer(s.expectedPeers[0].identifier, s.expectedPeers[0].subscribers[0])
+				return err
+			}
+			return
+		}(),
+		func() (s testStruct) {
+			s.msg = "one retain on release"
+			s.agent = NewDefaultAgent()
+			s.expectedPeers = []peerExpectation{}
+			s.appliedFunc = func(a *Agent) error {
+				pid := hostport.NewPeerIdentifier("localhost:1234")
+
+				sub := transporttest.NewMockPeerSubscriber(mockCtrl)
+
+				_, err := a.RetainPeer(pid, sub)
+				if err != nil {
+					return err
+				}
+				err = a.ReleasePeer(pid, sub)
+
+				return err
+			}
+			return
+		}(),
+		func() (s testStruct) {
+			s.msg = "one retain on release using peer"
+			s.agent = NewDefaultAgent()
+			s.expectedPeers = []peerExpectation{}
+			s.appliedFunc = func(a *Agent) error {
+				pid := hostport.NewPeerIdentifier("localhost:1234")
+
+				sub := transporttest.NewMockPeerSubscriber(mockCtrl)
+
+				peer, err := a.RetainPeer(pid, sub)
+				if err != nil {
+					return err
+				}
+				err = a.ReleasePeer(peer, sub)
+
+				return err
+			}
+			return
+		}(),
+		func() (s testStruct) {
+			s.msg = "three retains"
+			s.agent = NewDefaultAgent()
+			s.expectedPeers = createPeerExpectations(
+				mockCtrl,
+				[]string{"localhost:1234"},
+				3,
+			)
+			s.appliedFunc = func(a *Agent) error {
+				a.RetainPeer(s.expectedPeers[0].identifier, s.expectedPeers[0].subscribers[0])
+				a.RetainPeer(s.expectedPeers[0].identifier, s.expectedPeers[0].subscribers[1])
+				_, err := a.RetainPeer(s.expectedPeers[0].identifier, s.expectedPeers[0].subscribers[2])
+				return err
+			}
+			return
+		}(),
+		func() (s testStruct) {
+			s.msg = "three retains, one release"
+			s.agent = NewDefaultAgent()
+			s.expectedPeers = createPeerExpectations(
+				mockCtrl,
+				[]string{"localhost:1234"},
+				2,
+			)
+			s.appliedFunc = func(a *Agent) error {
+				unSub := transporttest.NewMockPeerSubscriber(mockCtrl)
```

Comment:
why not just call this 'subscriber' or 'sub'? unSub sounds counterintuitive, since it is, indeed a subscriber.